File: linda-core/src/main/java/it/unibo/coordination/linda/core/PredicativeTupleSpace.java
package it.unibo.coordination.linda.core;

import java.util.Optional;
import java.util.concurrent.CompletableFuture;

public interface PredicativeTupleSpace<T extends Tuple, TT extends Template, K, V> extends TupleSpace<T, TT, K, V> {
    CompletableFuture<Match<T, TT, K, V>> tryTake(TT template);

    default CompletableFuture<Optional<T>> tryTakeTuple(TT template) {
        return tryTake(template).thenApplyAsync(Match::getTuple);
    }

    CompletableFuture<Match<T, TT, K, V>> tryRead(TT template);

    default CompletableFuture<Optional<T>> tryReadTuple(TT template) {
        return tryRead(template).thenApplyAsync(Match::getTuple);
    }
}

File: linda-core/src/main/kotlin/it/unibo/coordination/linda/core/PredicativeTupleSpace.kt
package it.unibo.coordination.linda.core

import it.unibo.coordination.Promise
import java.util.*

interface PredicativeTupleSpace<T : Tuple, TT : Template, K, V, M : Match<T, TT, K, V>> : LindaTupleSpace<T, TT, K, V, M> {
    fun tryTake(template: TT): Promise<M>

    @JvmDefault
    fun tryTake(template: String): Promise<M> =
            tryTake(template.toTemplate())

    @JvmDefault
    fun tryTakeTuple(template: TT): Promise<Optional<T>> {
        return tryTake(template).thenApplyAsync{ it.tuple }
    }

    @JvmDefault
    fun tryTakeTuple(template: String): Promise<Optional<T>> =
            tryTakeTuple(template.toTemplate())

    fun tryRead(template: TT): Promise<M>

    @JvmDefault
    fun tryRead(template: String): Promise<M> =
            tryRead(template.toTemplate())

    @JvmDefault
    fun tryReadTuple(template: TT): Promise<Optional<T>> {
        return tryRead(template).thenApplyAsync{ it.tuple }
    }

    @JvmDefault
    fun tryReadTuple(template: String): Promise<Optional<T>> =
            tryReadTuple(template.toTemplate())
}
